src/qibo/models/encodings.py

```diff
@@ -1332,48 +1362,67 @@ def _intermediate_gate(
     return gate, lex_order, initial_string, phase_index
 
 
-def _sort_data_sparse(data, nqubits):
+def _sort_data_sparse(data, nqubits, backend):
     from qibo.quantum_info.utils import (  # pylint: disable=import-outside-toplevel
         hamming_weight,
     )
 
-    bitstrings, _data = [], []
-    for bitstring, elem in data:
-        if isinstance(bitstring, int) or "int" in str(type(bitstring)):
-            bitstring = f"{bitstring:0{nqubits}b}"
-        bitstrings.append(bitstring)
-        _data.append(elem)
-    _data = list(zip(bitstrings, _data))
+    # TODO: Fix this mess with qibo native data types
+    try:
+        test_dtype = bool("int" in str(data[0][0].dtype))
+    except AttributeError:
+        test_dtype = bool("int" in str(type(data[0][0])))
+
+    _data = [(f"{row[0]:0{nqubits}b}", row[1]) for row in data] if test_dtype else data
+
     _data = sorted(_data, key=lambda x: hamming_weight(x[0]))
 
-    data_sorted, bitstrings_sorted = [], []
-    for string, elem in _data:
-        bitstrings_sorted.append(np.array(list(string)).astype(int))
-        data_sorted.append(elem)
+    bitstrings_sorted, data_sorted = zip(*_data)
+    bitstrings_sorted = [
+        np.array(list(string)).astype(int) for string in bitstrings_sorted
+    ]
+
+    bitstrings_sorted = backend.cast(bitstrings_sorted, dtype=backend.np.int8)
+    data_sorted = backend.cast(data_sorted, dtype=data_sorted[0].dtype)
 
     return data_sorted, bitstrings_sorted
 
 
 def _get_gate_sparse(
-    distance, difference, touched_qubits, complex_data, controls, hw_0, hw_1, theta, phi
+    distance,
+    difference,
+    touched_qubits,
+    complex_data,
+    controls,
+    hw_0,
+    hw_1,
+    theta,
+    phi,
+    backend=None,
 ):
+    backend = _check_backend(backend)
     if distance == 1:
-        qubit = np.where(difference == 1)[0][0]
+        qubit = int(backend.np.where(difference == 1)[0][0])
         if qubit not in touched_qubits:
-            touched_qubits.append(qubit)
+            touched_qubits.append(int(qubit))
```

Comment:
```suggestion
            touched_qubits.append(qubit)
```
qubit should already be an int, as you just casted it

Reply:
They need to be actual `int`s instead of `cp.int8` otherwise some other errors are raised down the line.

Reply:
but you already did the explicit cast here https://github.com/qiboteam/qibo/blob/db5c529106537cc5ce944de344134b0d4f64fe3c/src/qibo/models/encodings.py#L1405

you're doing:
``` python
qubit = int(...)
if qubit not in touched_qubits:
    # cast again, but qubit is already an int
    touched_qubits.append(int(qubit))
```